File: 2812-find-the-safest-path-in-a-grid/2812-find-the-safest-path-in-a-grid.py
from collections import deque
from typing import List

class Solution:
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        n = len(grid)
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        # Step 1: Multi-Source BFS to compute distances
        dist = [[float('inf')] * n for _ in range(n)]
        q = deque()
        
        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    dist[i][j] = 0
                    q.append((i, j))
        
        while q:
            r, c = q.popleft()
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < n and dist[nr][nc] == float('inf'):
                    dist[nr][nc] = dist[r][c] + 1
                    q.append((nr, nc))
        
        # Step 2: Binary Search
        def can_reach(limit: int) -> bool:
            if dist[0][0] < limit:
                return False
            
            vis = [[False] * n for _ in range(n)]
            q = deque([(0, 0)])
            vis[0][0] = True
            
            while q:
                r, c = q.popleft()
                if r == n - 1 and c == n - 1:
                    return True
                
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < n and 0 <= nc < n and not vis[nr][nc] and dist[nr][nc] >= limit:
                        vis[nr][nc] = True
                        q.append((nr, nc))
            
            return False
        
        low, high, ans = 0, 2 * n, 0
        while low <= high:
            mid = (low + high) // 2
            if can_reach(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
        
        return ans

File: 2812-find-the-safest-path-in-a-grid/2812-find-the-safest-path-in-a-grid.rs
use std::collections::VecDeque;

impl Solution {
    pub fn maximum_safeness_factor(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let dirs = vec![(-1, 0), (1, 0), (0, -1), (0, 1)];
        
        // Step 1: Multi-Source BFS
        let mut dist = vec![vec![i32::MAX; n]; n];
        let mut q = VecDeque::new();
        
        for i in 0..n {
            for j in 0..n {
                if grid[i][j] == 1 {
                    dist[i][j] = 0;
                    q.push_back((i, j));
                }
            }
        }
        
        while let Some((r, c)) = q.pop_front() {
            for &(dr, dc) in &dirs {
                let nr = r as isize + dr;
                let nc = c as isize + dc;
                if nr >= 0 && nr < n as isize && nc >= 0 && nc < n as isize {
                    let nr = nr as usize;
                    let nc = nc as usize;
                    if dist[nr][nc] == i32::MAX {
                        dist[nr][nc] = dist[r][c] + 1;
                        q.push_back((nr, nc));
                    }
                }
            }
        }
        
        // Step 2: Binary Search
        let can_reach = |limit: i32| -> bool {
            if dist[0][0] < limit {
                return false;
            }
            
            let mut vis = vec![vec![false; n]; n];
            let mut q = VecDeque::new();
            q.push_back((0, 0));
            vis[0][0] = true;
            
            while let Some((r, c)) = q.pop_front() {
                if r == n - 1 && c == n - 1 {
                    return true;
                }
                
                for &(dr, dc) in &dirs {
                    let nr = r as isize + dr;
                    let nc = c as isize + dc;
                    if nr >= 0 && nr < n as isize && nc >= 0 && nc < n as isize {
                        let nr = nr as usize;
                        let nc = nc as usize;
                        if !vis[nr][nc] && dist[nr][nc] >= limit {
                            vis[nr][nc] = true;
                            q.push_back((nr, nc));
                        }
                    }
                }
            }
            
            false
        };
        
        let mut low = 0;
        let mut high = (2 * n) as i32;
        let mut ans = 0;
        
        while low <= high {
            let mid = low + (high - low) / 2;
            if can_reach(mid) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        ans
    }
}